src/proxy/cpsubsystem/BaseCPProxy.ts

```diff
@@ -15,35 +15,41 @@
  */
 /** @ignore *//** */
 
-import {HazelcastClient} from '../../HazelcastClient';
 import {ClientMessage} from '../../protocol/ClientMessage';
 import {RaftGroupId} from './RaftGroupId';
 import {CPGroupDestroyCPObjectCodec} from '../../codec/CPGroupDestroyCPObjectCodec';
 import {UnsupportedOperationError} from '../../core';
 import {Data} from '../../serialization/Data';
+import {SerializationService} from '../../serialization/SerializationService';
+import {InvocationService} from '../../invocation/InvocationService';
 
 /**
  * Common super class for any CP Subsystem proxy.
  * @internal
  */
 export abstract class BaseCPProxy {
 
-    protected client: HazelcastClient;
     protected readonly proxyName: string;
     protected readonly serviceName: string;
```

Comment:
My ide says that various class attributes(about 30 of them) can be done readonly. they are the ones that I did not add in this PR

Reply:
> My ide says that various class attributes(about 30 of them) can be done readonly. they are the ones that I did not add in this PR.

Feel free to mark them as `readonly` as a part of this PR.

> it says that some methods can be static. why are not they if they can be? Is it something good to make methods static when it can be done?

Some of methods could be converted to static, but static method invocation requires `ClassName.methodName()` syntax which makes the code slightly less readable. I usually prefer to use functions instead of static methods. Performance-wise static methods shouldn't make any difference.

Long story short, let's keep those methods as is, at least for now. This PR is already big enough.